File: packages/nexus-sdk-go/client_test.go
package nexus

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

// NOTE: written but not executed — see client.go's package doc comment
// for why (no Go toolchain in the environment this codebase was built
// in). Uses httptest.NewServer, the standard library's own real-local-
// server approach to testing an HTTP client, so it needs no mocking
// framework beyond net/http/httptest itself.

func TestGetCreditBalance_ReturnsDataPayload(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/nexus/api/v1/credit/balance" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}
		if r.Header.Get("Authorization") != "Bearer nx_test" {
			t.Errorf("unexpected Authorization header: %s", r.Header.Get("Authorization"))
		}
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]any{"data": map[string]any{"businessId": 1, "balance": 500}})
	}))
	defer server.Close()

	client := New(server.URL, "nx_test")
	result, err := client.GetCreditBalance()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["balance"] != float64(500) {
		t.Errorf("expected balance 500, got %v", result["balance"])
	}
}

func TestGetCatalog_PassesQueryParameter(t *testing.T) {
	var capturedQuery string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		capturedQuery = r.URL.RawQuery
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]any{"data": map[string]any{"products": []any{}, "services": []any{}}})
	}))
	defer server.Close()

	client := New(server.URL, "nx_test")
	if _, err := client.GetCatalog("widget"); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if capturedQuery != "query=widget" {
		t.Errorf("expected query=widget, got %s", capturedQuery)
	}
}

func TestErrorResponse_ReturnsAPIErrorWithEnvelopeDetails(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusForbidden)
		json.NewEncoder(w).Encode(map[string]any{"error": map[string]any{"code": "FORBIDDEN", "message": "Missing scope: catalog.read"}})
	}))
	defer server.Close()

	client := New(server.URL, "nx_test")
	_, err := client.GetCatalog("")

	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T", err)
	}
	if apiErr.HTTPStatus != http.StatusForbidden || apiErr.Code != "FORBIDDEN" || apiErr.Message != "Missing scope: catalog.read" {
		t.Errorf("unexpected APIError: %+v", apiErr)
	}
}

func TestVerifyWebhookSignature_ValidSignature_ReturnsTrue(t *testing.T) {
	body := []byte(`{"event":"negotiation.accepted"}`)
	secret := "shhh"
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(body)
	signature := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	if !VerifyWebhookSignature(body, signature, secret) {
		t.Error("expected valid signature to verify")
	}
}

func TestVerifyWebhookSignature_TamperedBody_ReturnsFalse(t *testing.T) {
	secret := "shhh"
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(`{"event":"original"}`))
	signature := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	if VerifyWebhookSignature([]byte(`{"event":"tampered"}`), signature, secret) {
		t.Error("expected tampered body to fail verification")
	}
}

File: packages/nexus-sdk-node/test/client.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { NexusClient, NexusApiError } from '../src/index.js';

function fakeResponse(status, body) {
    return {
        ok: status < 400,
        status,
        json: async () => body,
    };
}

test('getCreditBalance returns the data payload', async () => {
    const client = new NexusClient('https://nexus.example.com', 'nx_test', {
        fetchImpl: async (url, options) => {
            assert.equal(url, 'https://nexus.example.com/nexus/api/v1/credit/balance');
            assert.equal(options.method, 'GET');
            return fakeResponse(200, { data: { businessId: 1, balance: 500 } });
        },
    });

    const result = await client.getCreditBalance();
    assert.equal(result.balance, 500);
});

test('getCatalog passes the query parameter', async () => {
    let capturedUrl;
    const client = new NexusClient('https://nexus.example.com', 'nx_test', {
        fetchImpl: async (url) => {
            capturedUrl = url;
            return fakeResponse(200, { data: { products: [], services: [] } });
        },
    });

    await client.getCatalog('widget');
    assert.equal(capturedUrl, 'https://nexus.example.com/nexus/api/v1/catalog?query=widget');
});

test('error response throws NexusApiError with envelope details', async () => {
    const client = new NexusClient('https://nexus.example.com', 'nx_test', {
        fetchImpl: async () => fakeResponse(403, { error: { code: 'FORBIDDEN', message: 'Missing scope: catalog.read' } }),
    });

    await assert.rejects(
        () => client.getCatalog(),
        (err) => {
            assert.ok(err instanceof NexusApiError);
            assert.equal(err.httpStatus, 403);
            assert.equal(err.errorCode, 'FORBIDDEN');
            assert.equal(err.message, 'Missing scope: catalog.read');
            return true;
        },
    );
});

test('graphql sends query and variables as JSON body', async () => {
    let capturedBody;
    const client = new NexusClient('https://nexus.example.com', 'nx_test', {
        fetchImpl: async (url, options) => {
            capturedBody = JSON.parse(options.body);
            return fakeResponse(200, { data: { creditBalance: { balance: 10 } } });
        },
    });

    await client.graphql('{ creditBalance { balance } }', { foo: 'bar' });
    assert.equal(capturedBody.query, '{ creditBalance { balance } }');
    assert.deepEqual(capturedBody.variables, { foo: 'bar' });
});

test('verifyWebhookSignature accepts a valid signature', async () => {
    const { createHmac } = await import('node:crypto');
    const body = '{"event":"negotiation.accepted"}';
    const secret = 'shhh';
    const signature = `sha256=${createHmac('sha256', secret).update(body).digest('hex')}`;

    assert.equal(NexusClient.verifyWebhookSignature(body, signature, secret), true);
});

test('verifyWebhookSignature rejects a tampered body', async () => {
    const { createHmac } = await import('node:crypto');
    const secret = 'shhh';
    const signature = `sha256=${createHmac('sha256', secret).update('{"event":"original"}').digest('hex')}`;

    assert.equal(NexusClient.verifyWebhookSignature('{"event":"tampered"}', signature, secret), false);
});
